sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -400,13 +443,25 @@ private GatherResults(Coder<ResultT> resultCoder) {
     @Override
     public PCollection<List<ResultT>> expand(PCollection<ResultT> input) {
       if (getWindowedWrites()) {
-        // Reshuffle the results to make them stable against retries.
+        // Group the results to make them stable against retries.
         // Use a single void key to maximize size of bundles for finalization.
         return input
-            .apply("Add void key", WithKeys.of((Void) null))
-            .apply("Reshuffle", Reshuffle.of())
-            .apply("Drop key", Values.create())
-            .apply("Gather bundles", ParDo.of(new GatherBundlesPerWindowFn<>()))
+            .apply("AddVoidKey", WithKeys.of((Void) null))
```

Comment:
Two reasons:
- `Reshuffle` expands the iterable after GBK and "Gather bundles" effectively reverts the expansion by re-gathering the elements in a bundle. Why not just use GBK?
- `Reshuffle` didn't appear to work properly with auto-sharding where we might emit multiple outputs per window and 
`Reshuffle` didn't group those outputs in the same window (causing the added test to fail). My rough guess was that it might be due to that we added timestamp in the key to group on in `Reshuffle`:https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java#L87
but I could be entirely wrong.

One argument of restoring this change I can think of is that runners might do optimizations for `Reshuffle` and using `explicit` GBK will drop such optimization if any. To keep the original behavior I could refactor the code to only use explicit GBK in the new path. Any thoughts?

Reply:
After some investigation and discussions with @robertwb, the cause of `Reshuffle` not working for the new path is that it applies a `ReshuffleTrigger` that triggers on every element so we will get only one element per bundle downstream. So even though we call the following DoFn `GatherBundlesPerWindowFn`, it doesn't actually gather the elements per window. That works OK in the fixed sharding case since we encode the shard idx explicitly. However, in the auto-sharding case, we don't know the shard idx and have to rely on counting the shards within a bundle, meaning that we need to guarantee that the elements in the same window are gathered in the same bundle to perform the final renaming.